### PR TITLE
[Animal] Refresh code for comparing a ration's nutrient supply against an animal's nutrient demand

### LINKED ISSUE #2130: [Animal] Refresh the calculation of animal nutritive requirements
The methods for calculating animal nutritive requirements need to be updated in order for them to be compatible with the refreshed animal module.

## Status

The old calculations of animal requirements in `animal/ration/animal_requirements.py` have not been updated yet to work with the refreshed animal code in `biophysical`.

## Intention

Have the code that calculates animal requirements be written in a cleaner, more maintainable way which will utilize new data structures.

## Request

Rewrite the code that calculates animal nutritive requirements to work with the refreshed `Animal` class and submodules.

### PR DESCRIPTION
**Pull request by Isgotkowitz**
Refreshes code for comparing nutrition supply against nutrition demand.

### Context

Issue(s) closed by this pull request: closes #2130

### What

Compares the nutrients supplied by a ration against those required by animal. A flag indicates if the nutrients adequately supplied the demand, as well as a report containing all surpluses and deficiencies of nutrients that were supplied.

### Why

In order to compare a ration against an animal's requirements, the nutrients in the ration have to be compared against the animal's requirements in a very specific way. The result of this comparison decides whether the animal's outcomes (and requirements) need to be lowered, or if the ration needs to be reformulated.

### How

The evaluation functions were taken from the constraint functions in `ration_driver.py`, and placed in a dedicated class called `NutritionEvaluator`. This class takes in a `NutritionRequirements`, a `NutritionSupply`, and a flag indicating whether the animal being evaluated is a cow or not. After comparing the supply and requirements, a flag is returned indicating whether the supply is sufficient to meet the requirements and a report of the differences between the supply and demand.

### Test plan

The scaffolding of a unit test suite is provided, but will be filled in at later point. If there are any bugs in this code, they will be caught when the refreshed animal module undergoes heavy scrutiny.

### CODE REVIEW

File: RUFAS/biophysical/animal/data_types/nutrition_data_structures.py [hunk 1]
```diff
@@ -23,7 +23,9 @@ class NutritionRequirements:
     calcium : float
         Calcium requirement (g).
     phosphorus : float
-        Phosphorus requirement (g).
+        Phosphorus requirement calculated with either the NASEM or NRC methodologies (g).
+    secondary_phosphorus : float
+        Phosphorus requirement calculated with the dedicated animal phosphorus submodule (g).
```
Comment (Isgotkowitz) [inline RUFAS/biophysical/animal/data_types/nutrition_data_structures.py]: This dedicated phosphorus submodule is `biophysical/animal/nutrients/nutrients.py`. The phosphorus requirement from that module will be added into this data structure by `Animal`, and I will write up an issue to renamed the submodule and class.

File: tests/animal_module_tests/nutrient_tests/test_nutrition_evaluator.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    """Tests that calcium supplied in ration is correctly compared against an animal's calcium requirement."""
+    pass
+
+
+def test_check_phosphorus_supplied() -> None:
+    """Tests that phosphorus supplied in ration is correctly compared against an animal's phosphorus requirement."""
+    pass
+
+
+def test_check_protein_supplied() -> bool:
```
Comment (Isgotkowitz) [inline tests/animal_module_tests/nutrient_tests/test_nutrition_evaluator.py]:
```suggestion
def test_check_protein_supplied() -> None:
```

File: RUFAS/biophysical/animal/animal_module_constants.py [hunk 3]
```diff
@@ -61,8 +61,8 @@ class AnimalModuleConstants:
     MILK_LACTOSE: float = 4.85
     """Milk lactose content, percentage."""
 
-    DMI_CONSTRAINT_PERCENT: float = 0.20
-    """The +/- percentage of DMI estimated allowed for ration formulation"""
+    DMI_CONSTRAINT_FRACTION: float = 0.20
+    """The +/- fraction of DMI estimated allowed for ration formulation"""
```
Comment (ew3361zh) [inline RUFAS/biophysical/animal/animal_module_constants.py]:
```suggestion
    """The +/- fraction of DMI estimated allowed for ration formulation."""
```

File: RUFAS/biophysical/animal/animal_module_constants.py [hunk 4]
```diff
@@ -84,6 +84,15 @@ class AnimalModuleConstants:
     MINIMUM_CALCIUM: float = 0.0
     """Minimum calcium estimate, g/day"""
 
+    MINIMUM_NDF: float = 15.0
+    """Minimum percentage of neutral detergent fiber (NDF) that a ration's dry matter must be."""
```
Comment (ew3361zh) [inline RUFAS/biophysical/animal/animal_module_constants.py]:
```suggestion
    """Minimum percentage of a ration's dry matter that must be neutral detergent fiber (NDF)."""
```

File: RUFAS/biophysical/animal/animal_module_constants.py [hunk 5]
```diff
@@ -84,6 +84,15 @@ class AnimalModuleConstants:
...
+
+    MAXIMUM_NDF: float = 45.0
+    """Maximum percentage of neutral detergent fiber (NDF) that a ration's dry matter must be."""
```
Comment (ew3361zh) [inline RUFAS/biophysical/animal/animal_module_constants.py]:
```suggestion
    """Maximum percentage of a ration's dry matter that must be neutral detergent fiber (NDF)."""
```

File: RUFAS/biophysical/animal/animal_module_constants.py [hunk 6]
```diff
@@ -84,6 +84,15 @@ class AnimalModuleConstants:
...
+
+    MINIMUM_FAT: float = 7.0
+    """Minimum percentage of fat that a ration's dry matter must be."""
```
Comment (ew3361zh) [inline RUFAS/biophysical/animal/animal_module_constants.py]:
```suggestion
    """Minimum percentage of a ration's dry matter that must be fat."""
```

File: RUFAS/biophysical/animal/data_types/nutrition_data_structures.py (hunk 1 above)
Comment (ew3361zh) [inline RUFAS/biophysical/animal/data_types/nutrition_data_structures.py]:
```suggestion
        Phosphorus requirement calculated by the dedicated animal phosphorus submodule (g).
```

File: RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+            "phosphorus": cls._check_phosphorus_supplied,
+            "protein": cls._check_protein_supplied,
+            "ndf": cls._check_neutral_detergent_fiber_supplied,
+            "fat": cls._check_fat_content,
+            "dry_matter": cls._check_dry_matter_intake,
+        }
+        cow_energy_nutrition_checkers = heifer_energy_nutrition_checkers + {
+            "total_energy": cls._check_total_energy_supplied,
+            "lactation": cls._check_lactation_energy_supplied,
+        }
```
Comment (ew3361zh) [inline RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py]: Dictionaries can't be added together, I think you have to use the ** operator.

File: RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+
+        checkers = cow_energy_nutrition_checkers if is_cow else heifer_energy_nutrition_checkers
+        results = {name: method(requirements, supply) for name, method in checkers.items()}
+
+        evaluation = NutritionEvaluationResults(
+            total_energy=results.get("total_energy"),
+            maintenance_energy=results["maintentance"],
```
Comment (ew3361zh) [inline RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py]:
```suggestion
            maintenance_energy=results["maintenance"],
```
Which brings up another point - should these all just be results.get()s?

File: RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+
+        total_energy_requirement = sum(
+            [
+                requirements.lactation_energy,
+                requirements.growth_energy,
+                requirements.maintenance_energy,
+                requirements.activity_energy,
+                requirements.pregnancy_energy,
+            ]
+        )
```
Comment (ew3361zh) [inline RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py]: Are all of these different energy types guaranteed to use the same unit?

File: RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+            ndf_percent=results["ndf"],
+            fat_percent=results["fat"],
+            dry_matter=results["dry_matter"],
+        )
+
+        is_valid_ration = evaluation.is_valid_cow_ration if is_cow else evaluation.is_valid_heifer_ration
+        return is_valid_ration, evaluation
+
+    @classmethod
+    def _check_total_energy_supplied(requirements: NutritionRequirements, supply: NutritionSupply) -> float:
```
Comment (ew3361zh) [inline RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py]:
```suggestion
    def _calculate_total_energy_supplied(requirements: NutritionRequirements, supply: NutritionSupply) -> float:
```
Seems more like a calculation than a check (and matches the docstring definition better which I think is an accurate description).

File: RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+                requirements.maintenance_energy,
+                requirements.activity_energy,
+                requirements.pregnancy_energy,
+            ]
+        )
+
+        return energy_supplied - total_energy_requirement
+
+    @classmethod
+    def _check_activity_maintenance_energy_supplied(
```
Comment (ew3361zh) [inline RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py]:
```suggestion
    def _calculate_activity_maintenance_energy_supplied(
```

File: RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+        float
+            Difference between the maintenance energy supplied and the maintenance energy required (Mcal).
+
+        """
+        energy_requirement = requirements.activity_energy + requirements.maintenance_energy
+
+        return supply.maintenance_energy - energy_requirement
+
+    @classmethod
+    def _check_lactation_energy_supplied(requirements: NutritionRequirements, supply: NutritionSupply) -> float:
```
Comment (ew3361zh) [inline RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py]:
```suggestion
    def _calculate_lactation_energy_supplied(requirements: NutritionRequirements, supply: NutritionSupply) -> float:
```

File: RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+        float
+            Difference between the lactation energy supplied and the lactation energy required (Mcal).
+
+        """
+        energy_requirement = requirements.lactation_energy + requirements.pregnancy_energy
+
+        return supply.lactation_energy - energy_requirement
+
+    @classmethod
+    def _check_growth_energy_supplied(requirements: NutritionRequirements, supply: NutritionSupply) -> float:
```
Comment (ew3361zh) [inline RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py]:
```suggestion
    def _calculate_growth_energy_supplied(requirements: NutritionRequirements, supply: NutritionSupply) -> float:
```
et al

File: RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py [hunk 14]
```diff
@@ -0,0 +1,317 @@
...
+        float
+            Difference between the phosphorus supplied and the phosphorus required (g).
+
+        """
+        requirement = max(requirements.phosphorus, requirements.secondary_phosphorus)
+
+        return supply.phosphorus - requirement
+
+    @classmethod
+    def _check_protein_supplied(requirements: NutritionRequirements, supply: NutritionSupply) -> bool:
```
Comment (ew3361zh) [inline RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py]:
```suggestion
    def _calculate_protein_supplied(requirements: NutritionRequirements, supply: NutritionSupply) -> float:
```

File: RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py [hunk 15]
```diff
@@ -0,0 +1,317 @@
...
+        """
+        fat_percentage = supply.fat_supply / supply.dry_matter * GeneralConstants.FRACTION_TO_PERCENTAGE
+
+        return fat_percentage - AnimalModuleConstants.MINIMUM_FAT
+
+    @classmethod
+    def _check_dry_matter_intake(requirements: NutritionRequirements, supply: NutritionSupply) -> float:
+        """Checks that the dry matter supplied meets the requirement."""
+        """
+        Calculates amount by which supplied dry matter under- or overshoots the required amount dry matter.
```
Comment (ew3361zh) [inline RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py]: Multiple docstrings/function definitions

File: RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py [hunk 16]
```diff
@@ -0,0 +1,317 @@
...
+        ----------
+        _ : NutritionRequirements
+            This argument is provided to keep the method signature uniform with other helper methods.
+        supply : NutritionSupply
+            Energy and nutrition supplied by a ration.
+
+        Returns
+        -------
+        float
+            Percentage by which supplied NDF under- or overshoots the required NDF range.
```
Comment (ew3361zh) [inline RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py]: My preference here would be to include the unit for uniformity's sake and reword if the redundancy is annoying.

**Review by ew3361zh — COMMENTED** [review]
Nice work Ed, the modularity of this code compared to the old code is such a great step forward. I left some docstring/function name feedback and there's an issue in the main function with combining the dictionaries (95% sure).

Not sure if additional error handling is necessary and what here should be added to OM when it's up and running but it seems like a fair amount. Thoughts? Maybe that's better done at a later point?

File: RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py (hunk 7 above)
Comment (Isgotkowitz) [reply RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py]: Good catch! Would you have any opposition to me using the pipe (`|`) operator instead though?

File: RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py (hunk 8 above)
Comment (Isgotkowitz) [reply RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py]: Not necessarily. I deliberately used `.get` only for the attributes which are allowed to be `None` (total energy and lactation). I think this better reflects that these two attributes are different from the others.

File: RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py (hunk 9 above)
Comment (Isgotkowitz) [reply RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py]: They _should_ use the same unit, I think in this context it is ok to assume that they do.

Also I realized I wrote an `@property` method for total_energy_requirement, so I will replace this code with that instead

File: RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py (hunk 15 above)
Comment (Isgotkowitz) [reply RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py]: *Homer Simpson voice*: Doh!

File: RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py (hunk 16 above)
Comment (Isgotkowitz) [reply RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py]: Even with the units added in I think it reads alright

**Review by Isgotkowitz — COMMENTED** [review]
Nice catches Niko! Addressing comments now

File: RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py (hunk 7 above)
Comment (ew3361zh) [reply RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py]: totally fine by me.

File: RUFAS/biophysical/animal/data_types/nutrition_data_structures.py [hunk 17]
```diff
@@ -23,7 +23,9 @@ class NutritionRequirements:
     calcium : float
         Calcium requirement (g).
     phosphorus : float
-        Phosphorus requirement (g).
+        Phosphorus requirement calculated with either the NASEM or NRC methodologies (g).
+    secondary_phosphorus : float
```
Comment (JoeWaddell) [inline RUFAS/biophysical/animal/data_types/nutrition_data_structures.py]: Reminder to carry over changes made in: https://github.com/RuminantFarmSystems/MASM/pull/2146

File: RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py [hunk 18]
```diff
@@ -0,0 +1,306 @@
...
+
+        evaluation = NutritionEvaluationResults(
+            total_energy=results.get("total_energy"),
+            maintenance_energy=results["maintenance"],
+            lactation_energy=results.get("lactation"),
+            growth_energy=results["growth"],
+            calcium=results["calcium"],
+            phosphorus=results["phosphorus"],
+            metabolizable_protein=results["protein"],
+            ndf_percent=results["ndf"],
```
Comment (JoeWaddell) [inline RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py]: There needs to be a check for forage-specific NDF, see `forage_NDF_constraint` in `routines/animal/ration/ration_optimizer.py`

**Review by ew3361zh — APPROVED** [review]
Looks great, Ed. Tiny request to maybe update the test names to match the functions for now (even though they're not implemented)?

**Comment by Isgotkowitz** [comment]
> Looks great, Ed. Tiny request to maybe update the test names to match the functions for now (even though they're not implemented)?

Yes, good call. I totally forgot about thsoe

File: RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py [hunk 19]
```diff
@@ -0,0 +1,306 @@
...
+
+        """
+        heifer_energy_nutrition_checkers = {
+            "maintenance": cls._calculate_activity_maintenance_energy_supplied,
+            "growth": cls._calculate_growth_energy_supplied,
+            "calcium": cls._calculate_calcium_supplied,
+            "phosphorus": cls._calculate_phosphorus_supplied,
+            "protein": cls._calculate_protein_supplied,
+            "ndf": cls._calculate_neutral_detergent_fiber_supplied,
+            "fat": cls._calculate_fat_content,
```
Comment (JoeWaddell) [inline RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py]: For this and the dry matter check, should the `_supplied` suffix be used for consistency?

File: RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py [hunk 20]
```diff
@@ -0,0 +1,306 @@
...
+
+        """
+        fat_percentage = supply.fat_supply / supply.dry_matter * GeneralConstants.FRACTION_TO_PERCENTAGE
+
+        return fat_percentage - AnimalModuleConstants.MINIMUM_FAT
+
+    @classmethod
+    def _calculate_dry_matter_intake(requirements: NutritionRequirements, supply: NutritionSupply) -> float:
+        """
+        Calculates amount by which supplied dry matter under- or overshoots the required amount dry matter.
```
Comment (JoeWaddell) [inline RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py:312]: Should be fat, not phosphorus

File: RUFAS/biophysical/animal/animal_module_constants.py [hunk 21]
```diff
@@ -84,6 +84,15 @@ class AnimalModuleConstants:
     MINIMUM_CALCIUM: float = 0.0
     """Minimum calcium estimate, g/day"""
 
+    MINIMUM_NDF: float = 15.0
```
Comment (JoeWaddell) [inline RUFAS/biophysical/animal/animal_module_constants.py]: I think here it should be specified this is a minimum for the ration (pen level), and not requirement (animal level). I don't think we currently have overlap, but I could see a case in which a user could get confused thinking they're changing one and actually changing the other. Example would be if in the future we add a minimum protein requirement return, so individual animals have a requirement of at least X g protein, but the ration still could have a higher requirement (pen average/median/etc)

In short: adding a `_RATION` suffix, or otherwise indicating it's ration-specific would be a nice addition for clarity

File: RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py [hunk 22]
```diff
@@ -0,0 +1,306 @@
...
+    NutritionRequirements,
+    NutritionSupply,
+    NutritionEvaluationResults,
+)
+from RUFAS.general_constants import GeneralConstants
+
+
+class NutritionEvaluator:
+    """Checks if energy and nutrients supplied in a ration satisfy the demand for an individual animal."""
+
```
Comment (JoeWaddell) [inline RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py]: This docstring (and elsewhere as applicable) could be generalized to avoid specifying this method to only apply to an individual animal, since the way most farms feed wouldn't be on an individual animal basis but a pen average (the way we currently do it in the code is to check the ration against the average/mean/etc requirements for a pen).

**Review by JoeWaddell — COMMENTED** [review]
I really like how this is looking: going to be much easier for new users to read and interpret.

A few comments, but nothing major. Only outstanding concern is if the optimization methods can handle these returns (particularly that they're checking both mins and maxes simultaneously), but that's easy enough to handle when the optimizer is being addressed.

File: RUFAS/biophysical/animal/data_types/nutrition_data_structures.py (hunk 17 above)
Comment (Isgotkowitz) [reply RUFAS/biophysical/animal/data_types/nutrition_data_structures.py]: Thanks:)

File: RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py (hunk 19 above)
Comment (Isgotkowitz) [reply RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py]: Yes, good idea!

File: RUFAS/biophysical/animal/animal_module_constants.py (hunk 21 above)
Comment (Isgotkowitz) [reply RUFAS/biophysical/animal/animal_module_constants.py]: Is this the same for calcium and phosphorus too?

Comment (Isgotkowitz) [reply RUFAS/biophysical/animal/animal_module_constants.py]: And follow up question: if these are specific to the pen level and not the individual animal level, does that mean that a ration should be considered valid for an individual animal even if the NDF is outside the acceptable range for the pen level?

File: RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py (hunk 18 above)
Comment (Isgotkowitz) [reply RUFAS/biophysical/animal/nutrients/nutrition_evaluator.py]: Added! Good catch Joe, that was a big miss.

**Comment by Isgotkowitz** [comment]
Significant updates:
- Added calculation of the NDF supplied by forage in a ration.
- Added check for supply of forage NDF in the ration against minimum acceptable amount.
- Still need to verify how the check of nutrients differs when checking for an individual animal vs. checking for a pen average, and if further changes are needed to the code to accommodate any differences.
  - Update: there are differences between these, however the code does not have to be changed to accommodate these differences.

File: RUFAS/biophysical/animal/animal_module_constants.py (hunk 21 above)
Comment (JoeWaddell) [reply RUFAS/biophysical/animal/animal_module_constants.py]: We met via zoom, but to add some context here for future reference: the NDF check would apply to the individual or pen level (e.g. the former in cases of rations being formulated on an an individual basis).

for calcium and phorphorus, those are specific to individual animal requirement returns, not ration-level optimized/formulation checks.

**Comment by Isgotkowitz** [comment]
I'm not sure why there are 4 more errors in this branch, I checked them locally and the errors only appear to be in code that was added in other PRs.

File: RUFAS/biophysical/animal/data_types/nutrition_data_structures.py (hunk 1 above)
Comment (Isgotkowitz) [reply RUFAS/biophysical/animal/data_types/nutrition_data_structures.py]: Update: this other P requirement has been renamed to `process_based_phosphorus`.

**Review by JoeWaddell — APPROVED** [review]
Nice work handling the comments/changes, good progress here!

File: RUFAS/biophysical/animal/animal_module_constants.py [hunk 23]
```diff
@@ -84,6 +84,18 @@ class AnimalModuleConstants:
...
+    """Minimum percentage of a pen's ration's dry matter that must be neutral detergent fiber (NDF) (percent)."""
+
+    MAXIMUM_RATION_NDF: float = 45.0
+    """Maximum percentage of a pen's ration's dry matter that must be NDF (percent)."""
+
+    MINIMUM_RATION_FORAGE_NDF: float = 15.0
+    """Minimum percentage of a pen's ration's dry matter that must be NDF from forages (percent)."""
+
+    MINIMUM_RATION_FAT: float = 7.0
+    """Minimum percentage of a ration's dry matter that must be fat."""
```
Comment (Isgotkowitz) [inline RUFAS/biophysical/animal/animal_module_constants.py]:
```suggestion
    """Minimum percentage of a ration's dry matter that must be fat (percent)."""
```
